src/main/scala/mesosphere/marathon/state/PathId.scala

```diff
@@ -88,4 +91,43 @@ object PathId {
     def toPath: PathId = PathId(stringPath)
     def toRootPath: PathId = PathId(stringPath).canonicalPath()
   }
+
+  /**
+    * This regular expression is used to validate each path segment of an ID.
+    *
+    * If you change this, please also change "pathType" in AppDefinition.json and
+    * notify the maintainers of the DCOS CLI.
+    */
+  private[this] val ID_PATH_SEGMENT_PATTERN =
+    "^(([a-z0-9]|[a-z0-9][a-z0-9\\-]*[a-z0-9])\\.)*([a-z0-9]|[a-z0-9][a-z0-9\\-]*[a-z0-9])|(\\.|\\.\\.)$".r
+
+  implicit val pathIdValidator = validator[PathId] { pathId =>
+    pathId.path.each should matchRegexFully(ID_PATH_SEGMENT_PATTERN.pattern)
+    pathId is validChild
+  }
+
+  /**
+    * Check if path's parent, if it exists, is a valid parent indeed.
+    * @return
+    */
+  def validChild: Validator[PathId] = {
+    new Validator[PathId] {
+      override def apply(pathId: PathId): Result = {
+        if (pathId.parent == "".toPath) Success
```

Comment:
This change caused an integration test error. `PathId.empty` is not equal to `"".toPath`. The first produces a path id having `absolute = true`, the former a path id having `absolute = false`. I changed it back to `"".toPath` if you don't have another suggestion.